File: packages/core/src/rules/list.ts
import { isTag } from 'domhandler';
import type { Rule } from '../types.js';

export const listRules: Rule[] = [
    {
        filter: ['ul', 'ol'],
        replacement: ({ convertChildren, node, listDepth }) => {
            const content = convertChildren(node);
            // Top-level lists get surrounding newlines
            if (listDepth === 0) {
                return `\n\n${content.trim()}\n\n`;
            }
            // Nested lists are inline with parent li
            return `\n${content.trimEnd()}`;
        },
        priority: 0
    },

    {
        filter: 'li',
        replacement: ({ node, convertChildren, parent, options, listDepth, siblingIndex }) => {
            const content = convertChildren(node).trim();
            if (!content) return null;

            const indent = '  '.repeat(Math.max(0, listDepth - 1));
            const parentEl = parent && isTag(parent) ? parent : null;

            let bullet: string;
            if (parentEl && parentEl.name === 'ol') {
                const start = Number.parseInt(parentEl.attribs.start || '1', 10);
                // Count only <li> siblings before this one
                let liIndex = 0;
                for (let i = 0; i < siblingIndex; i++) {
                    const sibling = parentEl.children[i];
                    if (isTag(sibling) && sibling.name === 'li') {
                        liIndex++;
                    }
                }
                bullet = `${String(start + liIndex)}.`;
            } else {
                bullet = options.bulletChar;
            }

            // Handle multi-line content by indenting continuation lines
            const lines = content.split('\n');
            const first = `${indent}${bullet} ${lines[0]}`;
            if (lines.length === 1) return `${first}\n`;

            const continuationIndent = indent + ' '.repeat(bullet.length + 1);
            const rest = lines
                .slice(1)
                .map(line => (line.trim() ? `${continuationIndent}${line}` : ''))
                .join('\n');

            return `${first}\n${rest}\n`;
        },
        priority: 0
    }
];

File: python/src/markdown_for_agents/rules/list_.py
from __future__ import annotations

from .._types import Rule, RuleContext
from ..dom.nodes import Element, is_tag

list_rules: list[Rule] = [
    Rule(
        filter=["ul", "ol"],
        replacement=lambda ctx: _list_replacement(ctx),
        priority=0,
    ),
    Rule(
        filter="li",
        replacement=lambda ctx: _li_replacement(ctx),
        priority=0,
    ),
]


def _list_replacement(ctx: RuleContext) -> str:
    content = ctx.convert_children(ctx.node)
    if ctx.list_depth == 0:
        return f"\n\n{content.strip()}\n\n"
    return f"\n{content.rstrip()}"


def _li_replacement(ctx: RuleContext) -> str | None:
    content = ctx.convert_children(ctx.node).strip()
    if not content:
        return None

    indent = "  " * max(0, ctx.list_depth - 1)
    bullet = _get_bullet(ctx)

    lines = content.split("\n")
    first = f"{indent}{bullet} {lines[0]}"
    if len(lines) == 1:
        return f"{first}\n"

    continuation_indent = indent + " " * (len(bullet) + 1)
    rest_lines = [f"{continuation_indent}{line}" if line.strip() else "" for line in lines[1:]]

    return f"{first}\n{chr(10).join(rest_lines)}\n"


def _get_bullet(ctx: RuleContext) -> str:
    """Determine the bullet marker for a list item."""
    parent_el = ctx.parent if ctx.parent is not None and is_tag(ctx.parent) else None

    if not isinstance(parent_el, Element) or parent_el.name != "ol":
        return ctx.options.bullet_char

    start = int(parent_el.attribs.get("start", "1"))
    li_index = _count_preceding_li(parent_el, ctx.sibling_index)
    return f"{start + li_index}."


def _count_preceding_li(parent: Element, sibling_index: int) -> int:
    """Count <li> elements before the given sibling index."""
    count = 0
    for i in range(sibling_index):
        sibling = parent.children[i]
        if is_tag(sibling) and isinstance(sibling, Element) and sibling.name == "li":
            count += 1
    return count
